models/block.go

```diff
@@ -60,6 +78,72 @@ func decodeBlockEvent(event cadence.Event) (*Block, error) {
 			ReceiptRoot:         payload.ReceiptRoot,
 			TransactionHashRoot: payload.TransactionHashRoot,
 			TotalGasUsed:        payload.TotalGasUsed,
+			PrevRandao:          payload.PrevRandao,
 		},
 	}, nil
 }
+
+type blockV0 struct {
+	ParentBlockHash     gethCommon.Hash
+	Height              uint64
+	Timestamp           uint64
+	TotalSupply         *big.Int
+	ReceiptRoot         gethCommon.Hash
+	TransactionHashRoot gethCommon.Hash
+	TotalGasUsed        uint64
+}
+
+// Adds PrevRandao field
+type blockV1 struct {
+	ParentBlockHash     gethCommon.Hash
+	Height              uint64
+	Timestamp           uint64
+	TotalSupply         *big.Int
+	ReceiptRoot         gethCommon.Hash
+	TransactionHashRoot gethCommon.Hash
+	TotalGasUsed        uint64
+	PrevRandao          gethCommon.Hash
+}
+
+func pastBlockHash(b any) (gethCommon.Hash, error) {
+	data, err := rlp.EncodeToBytes(b)
+	return gethCrypto.Keccak256Hash(data), err
+}
+
+// decodeBlockBreakingChanges will try to decode the bytes into all
+// previous versions of block type, if it succeeds it will return the
+// migrated block, otherwise it will return nil.
+func decodeBlockBreakingChanges(encoded []byte) *Block {
```

Comment:
Isn't this a cop of the code from flow go?
should we just export those methods?

Reply:
Yeah, this is a copy from that code in flow-go. Though I am not sure if it still exists in the flow-go repo :thinking: .
We used to have these file changes for PreviewNet: https://github.com/onflow/flow-evm-gateway/commit/0e520ac78dc9d6a2853cffc58eb904228699f61a. But removed them for testnet launch.